paymentsheet/src/main/java/com/stripe/android/paymentsheet/PaymentSheetViewModel.kt

```diff
@@ -417,36 +420,91 @@ internal class PaymentSheetViewModel @Inject internal constructor(
+                    if (mode.isProcessingPayment) {
+                        PaymentIntentClientSecret(confirmResponse.clientSecret)
+                    } else {
+                        SetupIntentClientSecret(confirmResponse.clientSecret)
+                    }
+                }
             }
-        }
 
-        val confirmParamsFactory = ConfirmStripeIntentParamsFactory.createFactory(
-            clientSecret = clientSecret,
-            shipping = args.config?.shippingDetails?.toConfirmPaymentIntentShipping()
-        )
+            val confirmParamsFactory = ConfirmStripeIntentParamsFactory.createFactory(
+                clientSecret = clientSecret,
+                shipping = args.config?.shippingDetails?.toConfirmPaymentIntentShipping()
+            )
 
-        when (paymentSelection) {
-            is PaymentSelection.Saved -> {
-                confirmParamsFactory.create(paymentSelection)
+            when (paymentSelection) {
+                is PaymentSelection.Saved -> {
+                    confirmParamsFactory.create(paymentSelection)
+                }
+                is PaymentSelection.New -> {
+                    confirmParamsFactory.create(paymentSelection)
+                }
+                else -> null
+            }?.let { confirmParams ->
+                confirmStripeIntent(confirmParams)
             }
-            is PaymentSelection.New -> {
-                confirmParamsFactory.create(paymentSelection)
+        }
+    }
+
+    private suspend fun retrievePaymentMethodId(
+        paymentSelection: PaymentSelection?
+    ): String? {
+        return paymentSelection?.let {
+            when (paymentSelection) {
+                is PaymentSelection.Saved -> {
+                    paymentSelection.paymentMethod.id
+                }
+                is PaymentSelection.New -> {
+                    stripeRepository.createPaymentMethod(
+                        paymentSelection.paymentMethodCreateParams,
+                        ApiRequest.Options(
+                            apiKey = lazyPaymentConfig.get().publishableKey,
+                            stripeAccount = lazyPaymentConfig.get().stripeAccountId
+                        )
+                    )?.id
+                }
+                else -> {
+                    null
+                }
             }
-            else -> null
-        }?.let { confirmParams ->
-            confirmStripeIntent(confirmParams)
         }
     }
 
+    private suspend fun retrieveConfirmResponseForDeferredIntent(
+        paymentMethodId: String
+    ): ConfirmResponse {
+        return PaymentSheet.retrieveConfirmCallback?.onRetrieveConfirmResponse(
+            paymentMethodId
+        ) ?: throw IllegalStateException(
+            """
+                The DeferredIntent initialization mode requires the 
+                ClientSecretCallback to be implemented. Implement the callback using 
+                the PaymentSheet initializer.
+            """
```

Comment:
The fact that this illegal state is possible makes me a bit uneasy. On the other hand, we can’t move the `initializationMode` to the constructor as it will break existing integrations. Let’s chat a bit more about that next week!